lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp

```diff
@@ -9,25 +9,60 @@
 #include "PassDetails.h"
 #include "circt/Dialect/FIRRTL/FIRRTLAnnotations.h"
 #include "circt/Dialect/FIRRTL/FIRRTLAttributes.h"
+#include "circt/Dialect/FIRRTL/FIRRTLTypes.h"
 #include "circt/Dialect/FIRRTL/InstanceGraph.h"
 #include "circt/Dialect/FIRRTL/Passes.h"
+#include "circt/Support/LLVM.h"
 #include "mlir/IR/Threading.h"
 #include "llvm/ADT/APSInt.h"
+#include "llvm/ADT/STLExtras.h"
 #include "llvm/ADT/TinyPtrVector.h"
+#include "llvm/ADT/TypeSwitch.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "firrtl-imconstprop"
 
 using namespace circt;
 using namespace firrtl;
 
-/// Return true if this is a wire or register.
+/// Return true if this is either wire, register or similar(subfield and
+/// subindex).
 static bool isWireOrReg(Operation *op) {
-  return isa<WireOp>(op) || isa<RegResetOp>(op) || isa<RegOp>(op);
+  return isa<WireOp, RegResetOp, RegOp, SubfieldOp, SubindexOp>(op);
 }
 
 /// Return true if this is a wire or register we're allowed to delete.
 static bool isDeletableWireOrReg(Operation *op) {
   return isWireOrReg(op) && !hasDontTouch(op);
 }
 
+/// Return true if a type is possible to track. Currently, we allow passive
+/// types.
+static bool isTrackableType(Type type) {
+  return type.cast<FIRRTLType>().isPassive();
+}
+
+/// This function recursively apply `fn` to leaf ground types of `type`.
+static void traverseFIRRTLType(Type type,
+                               llvm::unique_function<void(FIRRTLType)> fn) {
```

Comment:
I think this should move up to FIRRTLTypes

Reply:
Thanks!